pkg/azure/client.go

```diff
@@ -118,50 +114,76 @@ func (az *azClient) SetAuthorizer(authorizer autorest.Authorizer) {
 	az.deploymentsClient.Authorizer = authorizer
 }
 
-func (az *azClient) GetGateway() (response n.ApplicationGateway, err error) {
-	err = utils.Retry(retryCount, retryPause,
+func (az *azClient) WaitForGetAccessOnGateway() (err error) {
+	glog.V(5).Info("Getting Application Gateway configuration.")
+	err = utils.Retry(-1, retryPause,
 		func() (utils.Retriable, error) {
-			response, err = az.appGatewaysClient.Get(az.ctx, string(az.resourceGroupName), string(az.appGwName))
+			response, err := az.appGatewaysClient.Get(az.ctx, string(az.resourceGroupName), string(az.appGwName))
 			if err == nil {
-				return utils.Retriable(false), nil
+				return utils.Retriable(true), nil
 			}
 
-			// Reasons for 403 errors
-			if response.Response.Response != nil && response.Response.StatusCode == 403 {
-				glog.Error("Following might be potential reasons:\n" +
-					" AKS Service Principal requires 'Managed Identity Operator' access on Controller Identity\n" +
-					" 'identityResourceID' and/or 'identityClientID' are incorrect in the Helm config\n" +
-					" AGIC Identity requires 'Contributor' access on Application Gateway and 'Reader' access on Application Gateway's Resource Group\n" +
-					" Please check the AAD Pod Identity mni and nmi pod logs to find potential issues.")
-			}
+			e := controllererrors.NewErrorWithInnerErrorf(
+				controllererrors.ErrorGetApplicationGatewayError,
+				err,
+				"Failed fetching configuration for Application Gateway. Will retry in %v.", retryPause,
+			)
 
-			if response.Response.Response != nil && response.Response.StatusCode == 404 {
-				err := controllererrors.NewErrorWithInnerError(
-					controllererrors.ErrorApplicationGatewayNotFound,
+			if response.Response.Response != nil {
+				e = controllererrors.NewErrorWithInnerErrorf(
+					controllererrors.ErrorApplicationGatewayUnexpectedStatusCode,
 					err,
-					"received 404 NOT FOUND status code on getting Application Gateway from ARM.",
+					"Unexpected status code '%d' while performing a GET on Application Gateway.", response.Response.StatusCode,
 				)
-				glog.Error(err.Error())
-				return utils.Retriable(false), err
+
+				if response.Response.StatusCode == 404 {
+					e.Code = controllererrors.ErrorApplicationGatewayNotFound
+				}
+
+				if response.Response.StatusCode == 403 {
+					e.Code = controllererrors.ErrorApplicationGatewayForbidden
+
+					clientID := "<agic-client-id>"
+					if az.clientID != "" {
+						clientID = az.clientID
+					}
+
+					e.Message += fmt.Sprintf(
+						" You can use 'az role assignment create --role Contributor --scope /subscriptions/%s/resourceGroups/%s --assignee %s' to assign permissions.",
```

Comment:
Should the scope be App-Gateway-ID instead of resource group? resource group only needs reader role

Reply:
1) Either we can add this:
```
az role assignment create --role Reader --scope /subscriptions/%s/resourceGroups/%s --assignee %s
az role assignment create --role Contributor --scope <appGatewayId> --assignee %s
```

2) Or we can aim for Contributor to the resource group
a) One command to give access
b) Might be helpful when vnet is in the same resource group. that way, we can use this access to apply route table to the subnet through AGIC without asking for addtitional permission to the Virtual network.

Reply:
I think we should go with 1st option, the role assignment should be granted as least as possible unless the contributor role becomes mandatory